### DISCUSSION TITLE
osu! Chaos Mode (spectator interaction with player)

### DISCUSSION THREAD

**Discussion by FPSwithaWacomTablet** [ppy/osu#15151] [Ideas]
#  Chaos Mode - The ability for spectators to change in-game settings in real-time

The direct analogy comes from Twitch chat playing Pokemon, where they could control the game using in-chat commands:

[![img](https://i.ytimg.com/an_webp/hEKbfId5oHM/mqdefault_6s.webp?du=3000&sqp=CLbqrYsG&rs=AOn4CLDJVkLw0J-Ey_UBlsK_3gxErOmulQ)](https://www.youtube.com/watch?v=hEKbfId5oHM)

Another analogy is from another streamer called KuruHS where his copy of NFS Most Wanted was modded to accept Twitch interactions to trigger in-game effects:

[![img](https://i.ytimg.com/vi/xMhbqcAvkCk/hq720.jpg?sqp=-oaymwEcCOgCEMoBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLDKP6CGhX9DhYoa8AOBdbwa3Zf0zQ)](https://www.youtube.com/watch?v=xMhbqcAvkCk)

## How this ties into osu?

Considering settings like tablet area and skins can be changed instantly in real-time, I would also assume that it also applies to mods as well. This shows promise when trying to create its osu equivalent.

## What I ask for a feature?

A simple in-game API for third-party bots and tools to change in-game settings outside of osu. Should be accessible using JS, Python, (insert popular web language here) that can be used by any competent *stack dev. ppy also mentioned that *maybe in-game spectators might also be able to tweak the settings as a mod* - quite spicy actually.

## Why I ask?

I tried changing the config file while the game was running and couldn't get it to change the settings live. Much sadness was experienced that day.

## Why is it here?

ppy told me to post it here after I pitched it to him. May the discussion ensue :)

**Comment by LeNitrous** [comment]
I think this is doable via #12274 since that PR allows exposes the game to a websocket. Just need to expand it to allow certain endpoints to control the game itself. Rather, it would be really helpful if osu! was ever put into an arcade cabinet now that you think about it.

**Reply by FPSwithaWacomTablet** [reply]
If someone could pinpoint the section of code where mods/settings are enabled/disabled/changed, this could get something started. Current progress with #12274 is also appreciated as that will help with the initial Websocket setup.

**Reply by peppy** [reply]
Please wait until that pull request is merged before attempting anything.

**Comment by marioman63** [comment]
I accomplished something like this with some external software for stream (sammiboard and autohotkey), where, if a twitch viewer redeemed a certain reward, it would enable a autohotkey script that would periodically send ctrl+R commands to the game, which would randomize my skin on the fly. was really fun, especially when it happens mid-song and you get some of the more meme-y skins in my collection. Would be cool if something like that could be done natively via some sort of integration. This proposal also makes me think of other potential, such as built in viewer song requests and the like.
